app/controllers/sessions_controller.rb

```diff
@@ -1,11 +1,40 @@
 class SessionsController < ApplicationController
+  skip_before_filter :verify_authenticity_token, only: :create
+  before_action :require_logged_out, except: [:destroy]
 
-  def new; end
+  MESSAGES = {
+    signout_success: "Peep ya later!"
+  }
+
+  def new
+    # Go to Instagram for authorization & confirmation
+    redirect_to Instagram.authorize_url(:redirect_uri => callback_url)
+  end
 
   def create
-    auth_hash = request.env['omniauth.auth'] 
-    # User.find_or_create_from_omniauth(auth_hash)
+    # Come back to our site
+    if params[:provider] == 'instagram'
+      response = Instagram.get_access_token(params[:code], :redirect_uri => callback_url)
+      session[:access_token] = response.access_token
+      user = User.find_or_create_from_omniauth(response) # Returns a user obj. or nil
+    else
+      auth_hash = request.env['omniauth.auth']
+      user = User.find_or_create_from_omniauth(auth_hash) # Returns a user obj. or nil
+    end
+
+    session[:user_id] = user.id
```

Comment:
While this works for ruby, it's a potential scoping risk to define a variable for the first time inside an block.

Normally, it's worth defining the variable before it goes into a block as a nil assignment: `user = nil`.

You could also take advantage of the way ruby processes return values and do something like this:
```
user = if params[:provider] == 'instagram'
  response = Instagram.get_access_token(params[:code], :redirect_uri => callback_url)
  session[:access_token] = response.access_token
  User.find_or_create_from_omniauth(response) # Returns a user obj. or nil
else
  auth_hash = request.env['omniauth.auth']
  User.find_or_create_from_omniauth(auth_hash) # Returns a user obj. or nil
end
```